### ISSUE TITLE
False positive in doc_markdown unbalanced backticks if code block in list contains backtick

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#7421]
```rust
//! - ```rust
//!   /// `lol`
//!   pub struct Struct;
//!   ```

fn main() {}
```

```console
$ cargo clippy --tests -- -Dclippy::pedantic
error: backticks are unbalanced
 --> src/main.rs:1:1
  |
1 | / //! - ```rust
2 | | //!   /// `lol`
  | |_______________^
  |
  = note: `-D clippy::doc-markdown` implied by `-D clippy::pedantic`
  = help: a backtick may be missing a pair
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_markdown
```

This seems like a false positive. There are no unbalanced backticks in the above.

<img src="https://user-images.githubusercontent.com/1940490/124341688-f4845500-db72-11eb-8750-721db1a5332b.png" width="500" alt="screenshot of rendered documentation">

Mentioning @ebobrow @xFrednet @flip1995 who touched this lint in #7357.

<br>

### Meta

- `cargo clippy -V`: clippy 0.1.55 (798baeb 2021-07-02)
- `rustc -Vv`:
  ```
  rustc 1.55.0-nightly (798baebde 2021-07-02)
  binary: rustc
  commit-hash: 798baebde1fe77e5a660490ec64e727a5d79970d
  commit-date: 2021-07-02
  host: x86_64-unknown-linux-gnu
  release: 1.55.0-nightly
  LLVM version: 12.0.1
  ```
